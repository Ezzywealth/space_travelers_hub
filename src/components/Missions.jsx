import React from 'react';
import { useDispatch } from 'react-redux';
import asyncMission from '../redux/api/apiSlice';

const Missions = () => {
  const dispatch = useDispatch();
  return (
    <button type="button" onClick={() => dispatch(asyncMission())}>
      click
    </button>
  );
};
export default Missions;
