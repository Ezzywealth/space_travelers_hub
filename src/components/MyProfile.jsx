import React from 'react';

const MyProfile = () => (
  <div>MyProfile MyProfile MyProfile MyProfile MyProfile MyProfile</div>
);

export default MyProfile;
