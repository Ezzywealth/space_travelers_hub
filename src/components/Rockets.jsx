import React from 'react';

const Rockets = () => (
  <div>
    Rockets Rockets Rockets Rockets Rockets Rockets Rockets Rockets Rockets
    Rockets Rockets Rockets Rockets Rockets Rockets
  </div>
);

export default Rockets;
