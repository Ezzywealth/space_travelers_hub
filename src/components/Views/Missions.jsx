import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import asyncMission from '../../redux/slices/api/apiSlice';
import Mission from './Mission';

const Missions = () => {
  const dispatch = useDispatch();
  const { missions } = useSelector((state) => state.missions);
  useEffect(() => {
    dispatch(asyncMission());
  }, [dispatch]);

  return (
    <>
      <table className="table-auto">
        <thead>
          <tr>
            <th className="border px-4 py-2">Mission</th>
            <th className="border px-4 py-2">Description</th>
            <th className="border px-4 py-2">Status</th>
            <th className="border px-4 py-2"> </th>
          </tr>
        </thead>
        <tbody>
          {missions.map((mission) => (
            <Mission
              key={mission.mission_id}
              missionId={mission.mission_id}
              missionName={mission.mission_name}
              description={mission.description}
              reserved={mission.reserved}
            />
          ))}
        </tbody>
      </table>
    </>
  );
};
export default Missions;
