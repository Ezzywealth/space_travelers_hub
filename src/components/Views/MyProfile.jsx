import React from 'react';
import { useSelector } from 'react-redux';

const MyProfile = () => {
  // const dispatch = useDispatch();
  const rockets = useSelector((state) => state.rockets.rockets);
  const { missions } = useSelector((state) => state.missions);
  return (
    <main className="grid md:grid-cols-2 gap-8 py-10 px-4 md:px-16">
      <article>
        <h2 className="text-[2rem] font-semibold mb-2 ">My Mission</h2>
        <section className="border border-[#121212]">
          {missions
            .filter((mission) => mission.reserved === true)
            .map((mission) => (
              <h4 className="p-3 border-b border-[#121212]" key={mission.mission_id}>
                {mission.mission_name}
              </h4>
            ))}
        </section>
      </article>
      <article>
        <h2 className="text-[2rem] font-semibold mb-2 ">My Rockets</h2>
        <section className="border border-[#121212]">
          {rockets
            .filter((rocket) => rocket.reserved === true)
            .map((rocket) => (
              <h4 className="p-3 border-b border-[#121212]" key={rocket.id}>
                {rocket.name}
              </h4>
            ))}
        </section>
      </article>
    </main>
  );
};

export default MyProfile;
