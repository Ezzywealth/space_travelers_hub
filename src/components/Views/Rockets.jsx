import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchRockets } from '../../redux/slices/rockets/rocketSlice';
import Rocket from '../Rocket';

const rocketsUrl = 'https://api.spacexdata.com/v3/rockets';
const Rockets = () => {
  const dispatch = useDispatch();
  const rockets = useSelector((state) => state.rockets.rockets);
  useEffect(() => {
    if (rockets.length < 1) {
      dispatch(fetchRockets(rocketsUrl));
    }
  }, [dispatch, rockets]);
  return (
    <main className="pb-10">
      <section className="px-4 md:px-20 pt-8 space-y-8 md:space-y-4">
        {rockets.map((rocket) => (
          <Rocket key={rocket.id} rocket={rocket} />
        ))}
      </section>
    </main>
  );
};

export default Rockets;
