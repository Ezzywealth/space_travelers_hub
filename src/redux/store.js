import { configureStore } from '@reduxjs/toolkit';
import rocketReducer from './slices/rockets/rocketSlice';
import missionReducer from './slices/missionSlice/missionSlice';
import menuReducer from './slices/menu/menuSlice';

const store = configureStore({
  reducer: {
    rockets: rocketReducer,
    missions: missionReducer,
    menu: menuReducer,
  },
});

export default store;
