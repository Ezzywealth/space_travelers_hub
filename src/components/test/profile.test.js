import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import TestRenderer, { act } from 'react-test-renderer';
import MyProfile from '../Views/MyProfile';
import store from '../../redux/store';
import Missions from '../Views/Missions';
import Rockets from '../Views/Rockets';

describe('Profile', () => {
  it('Snapshot of Profile Component', () => {
    const tree = TestRenderer.create(
      <Provider store={store}>
        <MyProfile />
        <Rockets />
        <Missions />
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Table of My mission exist', () => {
    render(
      <Provider store={store}>
        <MyProfile />
      </Provider>,
    );
    const tableElement = screen.getByText(/My Mission/i);
    expect(tableElement).toBeInTheDocument();
  });

  it('Table of My rockets exist', () => {
    render(
      <Provider store={store}>
        <Rockets />
        <Missions />
        <MyProfile />
      </Provider>,
    );
    const tableElement = screen.getByText(/My Rockets/i);
    expect(tableElement).toBeInTheDocument();
  });

  it('Check rockets buttons exist', async () => {
    render(
      <Provider store={store}>
        <MyProfile />
        <Rockets />
        <Missions />
      </Provider>,
    );
    const reserveBtn = await screen.findAllByText('Reserve Rocket');
    expect(reserveBtn).toHaveLength(4);
  });

  it('Reserve Rocket works properly', async () => {
    render(
      <Provider store={store}>
        <Rockets />
        <Missions />
        <MyProfile />
      </Provider>,
    );

    act(() => {
      const reserveBtn = screen.getAllByText('Reserve Rocket');
      reserveBtn[0].click();
    });

    const reckets = await screen.getAllByText('Reserve Rocket');
    expect(reckets).toHaveLength(3);
  });

  it('Check clicked rocket exist in Myprofile', async () => {
    render(
      <Provider store={store}>
        <MyProfile />
      </Provider>,
    );
    const rocketName = await screen.findAllByText('Falcon 1');
    expect(rocketName).toHaveLength(1);
  });

  it('cancel Rockets works properly', async () => {
    render(
      <Provider store={store}>
        <Rockets />
        <Missions />
        <MyProfile />
      </Provider>,
    );

    act(() => {
      const cancelBtns = screen.getAllByText('Cancel Reservation');
      cancelBtns[0].click();
    });
    const reserveBtns = await screen.getAllByText('Reserve Rocket');
    expect(reserveBtns).toHaveLength(4);
  });

  it('Check canceled rocket does not exist in Myprofile', async () => {
    render(
      <Provider store={store}>
        <MyProfile />
      </Provider>,
    );
    const rockets = await screen.getByTestId('rocket-container');
    expect(rockets.childElementCount).toBe(0);
  });

  it('Check missions buttons exist', async () => {
    render(
      <Provider store={store}>
        <Missions />
        <MyProfile />
      </Provider>,
    )
    const joinBtn = await screen.findAllByText('Join Mission');
    expect(joinBtn).toHaveLength(10);
  });

  it('Reserve Mission works properly', async () => {
    render(
      <Provider store={store}>
        <Rockets />
        <Missions />
        <MyProfile />
      </Provider>,
    );

    act(() => {
      const joinBtns = screen.getAllByText('Join Mission');
      joinBtns[0].click();
    });
    const status = await screen.getAllByText('Active Member');
    expect(status).toHaveLength(1);
  });

  it('Check clicked mission exist in Myprofile', async () => {
    render(
      <Provider store={store}>
        <MyProfile />
      </Provider>,
    );
    const missionName = await screen.findAllByText('Thaicom');
    expect(missionName).toHaveLength(1);
  });

  it('leave Mission works properly', async () => {
    render(
      <Provider store={store}>
        <Rockets />
        <Missions />
        <MyProfile />
      </Provider>,
    );

    act(() => {
      const leaveBtn = screen.getAllByText('Leave Mission');
      leaveBtn[0].click();
    });
    const status = await screen.getAllByText('NOT A MEMBER');
    expect(status).toHaveLength(10);
  });

  it('Check leaved mission does not exist in Myprofile', async () => {
    render(
      <Provider store={store}>
        <MyProfile />
      </Provider>,
    );
    const missions = await screen.getByTestId('mission-container');
    expect(missions.childElementCount).toBe(0);
  });

});
