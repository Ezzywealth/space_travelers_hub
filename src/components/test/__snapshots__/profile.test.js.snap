// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Snapshot of Profile Component 1`] = `
Array [
  <main
    className="grid md:grid-cols-2 gap-8 py-10 px-4 md:px-16"
  >
    <article>
      <h2
        className="text-[2rem] font-semibold mb-2 "
      >
        My Mission
      </h2>
      <section
        className="border border-[#121212]"
        data-testid="mission-container"
      />
    </article>
    <article>
      <h2
        className="text-[2rem] font-semibold mb-2 "
      >
        My Rockets
      </h2>
      <section
        className="border border-[#121212]"
        data-testid="rocket-container"
      />
    </article>
  </main>,
  <main
    className="pb-10"
  >
    <section
      className="px-4 md:px-20 pt-8 space-y-8 md:space-y-4"
    />
  </main>,
  <main
    className="overflow-scroll"
  >
    <table
      className="static table-auto mx-6"
    >
      <thead>
        <tr>
          <th
            className="border px-4 py-2"
          >
            Mission
          </th>
          <th
            className="border px-4 py-2"
          >
            Description
          </th>
          <th
            className="border px-4 py-2"
          >
            Status
          </th>
          <th
            className="border px-4 py-2"
          >
             
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </main>,
]
`;
